File: ios/Classes/Coin.swift
//
//  Coin.swift
//  Pods-Runner
//
//  Created by Jay on 3/29/22.
//
import WalletCore
import Flutter

class Coin: CoinProtocol {
    
    let name: String
    let coinType: CoinType
    
    init(name: String, coinType: CoinType){
        self.name = name
        self.coinType = coinType
    }
    
    func getName() -> String {
        return self.name
    }
    
    func generateAddress(path: String, mnemonic: String, passphrase: String) -> [String: String]? {
        let privateKey =  HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        if privateKey != nil {
            return  ["legacy": self.coinType.deriveAddress(privateKey: privateKey!)]
        }
        return nil
    }
    
    func getPrivateKey(path: String, mnemonic: String, passphrase: String) -> String? {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        let privateKey: String? = wallet!.getKey(coin: self.coinType, derivationPath: path).data.base64EncodedString()
        return privateKey
    }
    
    func getRawPrivateKey(path: String, mnemonic: String, passphrase: String) -> [UInt8] {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        return wallet!.getKey(coin: self.coinType, derivationPath: path).data.bytes
    }
   
    func getSeed(path: String, mnemonic: String, passphrase: String) -> [UInt8] {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        return wallet!.seed.bytes
    }
    
    func getPublicKey(path: String, mnemonic: String, passphrase: String) -> String? {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        let publicKey: String? = wallet!.getKey(coin: self.coinType, derivationPath: path).getPublicKeySecp256k1(compressed: true).data.base64EncodedString()
        return publicKey
    }
    
    func getRawPublicKey(path: String, mnemonic: String, passphrase: String) -> [UInt8] {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        return wallet!.getKey(coin: self.coinType, derivationPath: path).getPublicKeySecp256k1(compressed: true).data.bytes
    }
    
    func validateAddress(address: String) -> Bool {
        return self.coinType.validate(address: address)
    }

    func signTransaction(path: String, txData: [String: Any], mnemonic: String, passphrase: String) -> String? {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        let privateKey = wallet?.getKey(coin: self.coinType, derivationPath: path)
        let opJson =  Utils.objToJson(from: txData)
        if privateKey != nil {
            return AnySigner.signJSON(opJson!, key: privateKey!.data, coin: self.coinType)
        }
        return nil
    }
}

File: android/src/main/kotlin/africa/ejara/trustdart/Coin.kt
package africa.ejara.trustdart

import africa.ejara.trustdart.interfaces.CoinInterface
import africa.ejara.trustdart.utils.base64String
import africa.ejara.trustdart.utils.toHex
import africa.ejara.trustdart.utils.toHexByteArray
import org.json.JSONObject
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet

open class Coin<T1, T2>(nameOfCoin: T1, typeOfCoin: T2) : CoinInterface {

    var name: T1? = null
    var coinType: T2? = null

    init {
        name = nameOfCoin
        coinType = typeOfCoin
    }

    override fun generateAddress(
        path: String,
        mnemonic: String,
        passphrase: String
    ): Map<String, String>? {
        val wallet = HDWallet(mnemonic, passphrase)
        return mapOf("legacy" to (coinType!! as CoinType).deriveAddress(wallet.getKey(coinType as CoinType?, path)))
    }

    override fun getPrivateKey(path: String, mnemonic: String, passphrase: String): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType as CoinType?, path).data().base64String()
    }

    override fun getSeed(path: String, mnemonic: String, passphrase: String): ByteArray? {
        val data = HDWallet(mnemonic, passphrase).seed()
        return data
    }

    override fun getRawPrivateKey(path: String, mnemonic: String, passphrase: String): ByteArray? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType as CoinType?, path).data()
    } 

    override fun getPublicKey(path: String, mnemonic: String, passphrase: String): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType as CoinType?, path).getPublicKeySecp256k1(true).data().base64String()
    }
 
    override fun getRawPublicKey(path: String, mnemonic: String, passphrase: String): ByteArray? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType as CoinType?, path).getPublicKeySecp256k1(true).data()
    }

    override fun validateAddress(address: String): Boolean {
        return (coinType!! as CoinType).validate(address)
    }

    override fun signDataWithPrivateKey(
        path: String,
        mnemonic: String,
        passphrase: String,
        txData: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType as CoinType?, path)
        return privateKey.sign(txData.toHexByteArray(), (coinType!! as CoinType).curve()).toHex()
    }

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType as CoinType?, path)
        val opJson = JSONObject(txData).toString()
        return AnySigner.signJSON(opJson, privateKey.data(), (coinType!! as CoinType).value())
    }

    override fun multiSignTransaction(
        txData: Map<String, Any>,
        privateKeys: ArrayList<String>
    ): String? {
        val opJson = JSONObject(txData).toString()
        val signatures = mutableListOf<String>()

        for (privateKey in privateKeys) {
            val signature = AnySigner.signJSON(opJson, privateKey.toByteArray(), (coinType!! as CoinType).value())
            signatures.add(signature)
        }
        return signatures.joinToString(",")
    }

}
